lxd/project/permissions.go

```diff
@@ -1514,3 +1514,113 @@ func AllowSnapshotCreation(p *api.Project) error {
+func AllowClusterGroup(p *api.Project, groupName string) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	// Skip the check if the project is not restricted
+	if shared.IsFalseOrEmpty(p.Config["restricted"]) {
+		return nil
+	}
+
+	if len(clusterGroupsAllowed) > 0 && !shared.StringInSlice(groupName, clusterGroupsAllowed) {
+		return fmt.Errorf("Project isn't allowed to use this cluster group: %q", groupName)
+	}
+
+	return nil
+}
+
+// CheckTargetMember checks if the given member is allowed for the project.
+// If nil it also returns the resolved target member node information.
+func CheckTargetMember(p *api.Project, targetMember string, allMembers []db.NodeInfo) (*db.NodeInfo, error) {
+	// Find target member.
+	for _, potentialMember := range allMembers {
+		if potentialMember.Name == targetMember {
+			// If restricted groups are specified then check member is in at least one of them.
+			err := AllowClusterMember(p, &potentialMember)
+			if err != nil {
+				return nil, api.StatusErrorf(http.StatusForbidden, err.Error())
+			}
+
+			return &potentialMember, nil
+		}
+	}
+
+	return nil, api.StatusErrorf(http.StatusNotFound, "Cluster member %q not found", targetMember)
+}
+
+// CheckTargetGroup checks if the given group is allowed for the project.
+func CheckTargetGroup(ctx context.Context, tx *db.ClusterTx, p *api.Project, groupName string) error {
+	// If restricted groups are specified then check the requested group is in the list.
+	err := AllowClusterGroup(p, groupName)
+	if err != nil {
+		return api.StatusErrorf(http.StatusForbidden, err.Error())
+	}
+
+	// Check if the target group exists.
+	targetGroupExists, err := cluster.ClusterGroupExists(ctx, tx.Tx(), groupName)
+	if err != nil {
+		return err
+	}
+
+	if !targetGroupExists {
+		return api.StatusErrorf(http.StatusBadRequest, "Cluster group %q doesn't exist", groupName)
+	}
+
+	return nil
+}
+
+// CheckTarget checks if the given cluster target (member or group) is allowed.
+// If target is a cluster member, it returns the resolved node information and cluster group.
+// If target is a cluster group, it returns nil and the cluster group.
+// In case of error, neither node information nor cluster group gets returned.
+func CheckTarget(ctx context.Context, r *http.Request, tx *db.ClusterTx, p *api.Project, target string, allMembers []db.NodeInfo) (*db.NodeInfo, string, error) {
+	targetMember, targetGroup := shared.TargetDetect(target)
+
+	// Check manual cluster member targeting restrictions.
+	err := CheckClusterTargetRestriction(r, p, target)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if targetMember != "" {
+		member, err := CheckTargetMember(p, targetMember, allMembers)
+		if err != nil {
+			return nil, "", err
+		}
+
+		return member, targetGroup, nil
+	} else if targetGroup != "" {
+		err := CheckTargetGroup(ctx, tx, p, targetGroup)
+		if err != nil {
+			return nil, "", err
+		}
+
+		return nil, targetGroup, nil
+	}
+
+	return nil, targetGroup, nil
```

Comment:
This might be reached if `CheckTarget` is executed in a non clustered environment. For both the instance and instances endpoint this last return isn't required since `CheckTarget` is only called if `target != "" && clustered`.

Returning the empty `targetGroup` variable isn't very clear so I have changed it to `""`.